src/fa.c

```diff
@@ -4513,6 +4513,88 @@ void fa_dot(FILE *out, struct fa *fa) {
     fprintf(out, "}\n");
 }
 
+void fa_json(FILE *out, struct fa *fa) {
+    int *list_hashes;
+    int list_size = 100;
+    int num_states = 0;
+    int it;
+    char first = true;
+
+    fprintf(out,"{\n\t\"final\": [");
+
+    list_hashes = (int *)malloc(sizeof(int) * list_size);
```

Comment:
If I understand correctly, `F(ALLOC_N(list_hashes, list_size));` will automatically jump to the `error:` label in case of failure. Therefore, when restoring the hashes we need to check that we are not restoring any "false hash" (non-initialized position in the array). I've added the check `if (it == num_states) break;` in the loop to solve this issue.

Reply:
Yes, good catch - I've turned that into an if statement around the loop, but you are absolutely right that we need to make sure we do not try to switch hashes back in if we never changed them.